fontir/src/glyph.rs

```diff
@@ -181,89 +181,46 @@ enum GlyphOp {
 }
 
 /// Fix glyphs with mixed components/contours.
-///
-/// If there are component relationships between these glyphs the leaves have
-/// to be handled first.
+/// 
+/// We presume component cycles are checked elsewhere and do not check for them here
 fn resolve_inconsistencies(
     context: &Context,
-    todo: Vec<(GlyphOp, Arc<Glyph>)>,
+    mut todo: VecDeque<(GlyphOp, Arc<Glyph>)>,
     order: &mut GlyphOrder,
 ) -> Result<(), BadGlyph> {
-    let items = todo
-        .into_iter()
-        .map(|(op, glyph)| (glyph.name.clone(), (op, glyph)))
-        .collect::<HashMap<_, _>>();
+    let mut pending = todo
+        .iter()
+        .map(|(_, g)| g.name.clone())
+        .collect::<HashSet<_>>();
+    let mut curr_components = Vec::with_capacity(8); // avoid inner reallocs
+    'next_todo: while let Some((op, glyph)) = todo.pop_front() {
+        // Only fix curr if nothing else that needs fixing is reachable
+        curr_components.clear();
+        curr_components.extend(glyph.component_names().cloned());
+        while let Some(component_name) = curr_components.pop() {
+            if pending.contains(&component_name) {
+                // We can't got yet
+                todo.push_back((op, glyph));
+                continue 'next_todo;
+            }
+            let glyph = context.glyphs.get(&WorkId::Glyph(component_name));
+            curr_components.extend(glyph.component_names().cloned());
+        }
 
-    for name in component_ordering_within_glyphs(items.values().map(|(_, g)| g.clone())) {
-        let (op, glyph) = items.get(&name).unwrap();
+        // Our component graph doesn't reach any pending component, we are a go!
+        debug!("{op:?} {}", glyph.name);
         match op {
-            GlyphOp::ConvertToContour => convert_components_to_contours(context, glyph)?,
+            GlyphOp::ConvertToContour => convert_components_to_contours(context, &glyph)?,
             GlyphOp::MoveContoursToComponent => {
-                move_contours_to_new_component(context, order, glyph)?
+                move_contours_to_new_component(context, order, &glyph)?
             }
         }
-    }
-    Ok(())
-}
-
-// given a sequence of glyphs, return an ordering such that any glyph that
-// is a component of another glyph in the input set occurs before it.
-//
-// this has a funny signature so I can unit test it.
-fn component_ordering_within_glyphs(
```

Comment:
fwiw the only real reason that this existed as a separate fn was so that I could write a unit test against it. 🤷

Reply:
Added back one of the tests